src/CommandDialog.cpp

```diff
@@ -0,0 +1,161 @@
+#include "ScopedAssign.h"
+#include "CommandDialog.h"
+#include <QStandardItemModel>
+#include <map>
+
+
+enum CommandColumns {
+    ICON = 0, NAME, DESCRIPTION, COMMAND, HIDDEN, INDEX, MAX
+};
+
+
+CommandDialog::CommandDialog(QMap<QString, CommandRef>& commands, QWidget* parent)
+    : QDialog(parent), commands(commands)
+{
+    setupUi(this);
+
+    twCommands->resizeColumnsToContents();
+    twCommands->setColumnHidden(HIDDEN, true);
+    twCommands->verticalHeader()->hide();
+
+    connect(btnAdd, &QPushButton::clicked, this, &CommandDialog::onAddButtonClicked);
+    connect(btnRemove, &QPushButton::clicked, this, &CommandDialog::onRemoveButtonClicked);
+    connect(twCommands, &QTableWidget::itemChanged, this, &CommandDialog::changeTableItem);
+
+    connect(okButton, &QPushButton::clicked, this, &CommandDialog::accept);
+    connect(cancelButton, &QPushButton::clicked, this, &CommandDialog::reject);
+
+    refresh();
+}
+
+void CommandDialog::refresh()
+{
+    QList<QPair<CommandRef, QString>> list;
+
+    // set a comparison list
+    for (auto key: commands.keys()) {
+        list.append(QPair<CommandRef, QString>(commands[key], key));
+    }
+
+    // sort by index
+    std::sort(std::begin(list), std::end(list));
+    auto sa = ScopedAssign(userMode, false);
+
+    // populate
+    for (auto& pair: list) {
+        createItem();
+        writeCommand(pair.first);
+    }
+}
+
+void CommandDialog::onAddButtonClicked()
+{
+    auto sa = ScopedAssign(userMode, false);
+    createCommand();
+}
+
+void CommandDialog::onRemoveButtonClicked()
+{
+    auto sa = ScopedAssign(userMode, false);
+}
+
+int CommandDialog::createItem()
+{
+    int row = twCommands->rowCount();
+    twCommands->setRowCount(row + 1);
+    twCommands->selectRow(row);
+
+    auto* item0 = new QTableWidgetItem();
+    item0->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled | Qt::ItemIsEditable);
+    twCommands->setItem(row, ICON, item0);
+
+    auto* item1 = new QTableWidgetItem();
+    item1->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled | Qt::ItemIsEditable);
+    twCommands->setItem(row, NAME, item1);
+
+    auto* item2 = new QTableWidgetItem();
+    item2->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled | Qt::ItemIsEditable);
+    twCommands->setItem(row, DESCRIPTION, item2);
+
+    auto* item3 = new QTableWidgetItem();
+    item3->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled | Qt::ItemIsEditable);
+    twCommands->setItem(row, COMMAND, item3);
+
+    auto* item4 = new QTableWidgetItem();
+    twCommands->setItem(row, HIDDEN, item4);
```

Comment:
When possible I prefer to avoid such repeated code blocks (I've seen too many copy/paste bugs).
I think it might be possible like this:
```
for (auto column : {ICON, NAME, DESCRIPTION, COMMAND, HIDDEN}) {
    auto* item = new QTableWidgetItem();
    if (column != HIDDEN) {
        item->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled | Qt::ItemIsEditable);
    }
    twCommands->setItem(row, column, item);
}
```
```